### PR TITLE
Events

### PR DESCRIPTION
**Pull request by kobergj**
Brings events to reva!

Includes examples in `pkg/events/example`, Try them out!
Includes possibility to start a nats (stan) server

Will have no further effects as no services are configured to use eventing for now.

### CODE REVIEW

**Comment by labkode** [comment]
@kobergj looks promising, feel free to push against master as well, this functionality is more than welcome!

File: internal/grpc/interceptors/eventsmiddleware/events.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+		var ev interface{}
+		switch v := res.(type) {
+		case *collaboration.CreateShareResponse:
+			ev = ShareCreated(v)
+		}
+
+		if ev != nil {
+			if err := events.Publish(ev, stream); err != nil {
+				// TODO: should we error here? log? panic?
```
Comment (butonic) [inline internal/grpc/interceptors/eventsmiddleware/events.go:44]: panicing won't solve anything, the request has already been processed. emitting events is a post fact. all we can do is log an error and continue. The admin has to monitor these errors.

File: internal/grpc/interceptors/eventsmiddleware/events.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		return res, nil
+	}
+	return interceptor, defaultPriority, nil
+}
+
+// NewStream returns a new server stream interceptor
+// that creates the application context.
+func NewStream() grpc.StreamServerInterceptor {
+	interceptor := func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		// TODO: How to create events when streaming?
```
Comment (butonic) [inline internal/grpc/interceptors/eventsmiddleware/events.go:58]: ss has `SendMsg(m interface{})` and `RcvMsg(m interface{})`. The messages can be typecast as in the Unary interceptor.

File: internal/grpc/interceptors/eventsmiddleware/conversion.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import (
+	collaboration "github.com/cs3org/go-cs3apis/cs3/sharing/collaboration/v1beta1"
+	"github.com/cs3org/reva/pkg/events"
+)
+
+// ShareCreated converts response to event
+func ShareCreated(r *collaboration.CreateShareResponse) events.ShareCreated {
+	return events.ShareCreated{
+		Sharer: r.Share.Creator,
+		//Sharee: r.Share.Grantee, // TODO: unmarshaling fails -> find out why
```
Comment (butonic) [inline internal/grpc/interceptors/eventsmiddleware/conversion.go:12]: the micro nats implementation uses `"encoding/json"` not `jsonpb`, also see https://github.com/asim/go-micro/issues/237

File: internal/grpc/interceptors/eventsmiddleware/events.go (hunk 1 above)
Comment (kobergj) [reply internal/grpc/interceptors/eventsmiddleware/events.go]: Removed the comment

File: internal/grpc/interceptors/eventsmiddleware/events.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	defaultPriority = 200
+)
+
+func init() {
+	rgrpc.RegisterUnaryInterceptor("eventsmiddleware", NewUnary)
+}
+
+// NewUnary returns a new unary interceptor that emits events when needed
+func NewUnary(m map[string]interface{}) (grpc.UnaryServerInterceptor, int, error) {
+	stream, err := server.NewNatsStream()
```
[the remark this answers is not in the document]
Comment (kobergj) [reply internal/grpc/interceptors/eventsmiddleware/events.go]: Done as suggested

**Review by C0rby — APPROVED** [review]
LGTM, the only thing I would change is the logging. We should use the same logging framework as in the rest of the code but this can be changed later.

**Pull request merged**